src/cpp/session/SessionSourceDatabase.cpp

```diff
@@ -69,6 +69,10 @@ const char * const SourceDocument::SourceDocumentTypeCpp       = "cpp";
 
 namespace {
 
+// cached mapping of document id to document path (facilitates efficient path
+// lookup)
+std::map<std::string, std::string> s_idToPath;
```

Comment:
We also do this in SessionClang here: https://github.com/rstudio/rstudio/blob/master/src/cpp/session/modules/clang/SessionClang.cpp#L268-L280. High time we centralize this as @kevinushey suggested!